### PR TITLE
Cody for JetBrains: Add settings UI

### PR DESCRIPTION
**Pull request by vdavid**
- Closes https://github.com/sourcegraph/sourcegraph/issues/51999

Added settings UI (99% of this PR)

Also:
- Added "synchronized" modifier at two places
- Commented out a debug log line

Comprehensive, [2.5-min Loom](https://www.loom.com/share/93692ba3459546608df5d90b533205f6)

Screenshot:
CleanShot 2023-05-17 at 16 01 53@2x

## Test plan

- Tested in Loom
- Low stakes, not in production yet

### CODE REVIEW

**Comment by philipp-spiess** [comment]
The contrast here looks bad:

Screenshot 2023-05-17 at 18 26 51

**Comment by vdavid** [comment]
> The contrast here looks bad

Absolutely, I know. But it's only when the field is disabled, plus we have the same issue in the Sourcegraph plugin setting 😬, so I'd say it's fine for this MVP.

File: client/cody-jetbrains/src/main/java/com/sourcegraph/cody/config/NotificationActivity.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.startup.StartupActivity;
+import com.sourcegraph.cody.Icons;
+import org.jetbrains.annotations.NotNull;
+
+public class NotificationActivity implements StartupActivity.DumbAware {
+    @Override
+    public void runActivity(@NotNull Project project) {
+        String url = ConfigUtil.getEnterpriseUrl(project);
+        if (!ConfigUtil.areChatPredictionsEnabled(project) && (url.length() == 0 || url.startsWith(ConfigUtil.DOTCOM_URL))) {
```
Comment (philipp-spiess) [inline client/cody-jetbrains/src/main/java/com/sourcegraph/cody/config/NotificationActivity.java:18]: Why is `areChatPredictionsEnabled` in the condition here?

File: client/cody-jetbrains/src/main/java/com/sourcegraph/cody/config/SettingsConfigurableHelper.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        return !newSettings.getInstanceType().toString().equals(oldSettings.getInstanceType() != null ? oldSettings.getInstanceType() : SettingsComponent.InstanceType.DOTCOM.toString())
+            || !(newSettings.getDotcomAccessToken().equals(oldSettings.getDotcomAccessToken()) || newSettings.getDotcomAccessToken().isEmpty() && oldSettings.getDotcomAccessToken() == null)
+            || !newSettings.getEnterpriseUrl().equals(oldSettings.getEnterpriseUrl() != null ? oldSettings.getEnterpriseUrl() : "")
+            || !(newSettings.getEnterpriseAccessToken().equals(oldSettings.getEnterpriseAccessToken()) || newSettings.getEnterpriseAccessToken().isEmpty() && oldSettings.getEnterpriseAccessToken() == null)
+            || !newSettings.getCustomRequestHeaders().equals(oldSettings.getCustomRequestHeaders() != null ? oldSettings.getCustomRequestHeaders() : "")
+            || !newSettings.getCodebase().equals(oldSettings.getCodebase() != null ? oldSettings.getCodebase() : "")
+            || newSettings.areChatPredictionsEnabled() != Boolean.TRUE.equals(oldSettings.areChatPredictionsEnabled());
```
Comment (philipp-spiess) [inline client/cody-jetbrains/src/main/java/com/sourcegraph/cody/config/SettingsConfigurableHelper.java:21]: 😭

**Review by philipp-spiess — APPROVED** [review]
(insert-random-complaint-about-java)

File: client/cody-jetbrains/src/main/java/com/sourcegraph/cody/config/NotificationActivity.java (hunk 1 above)
Comment (vdavid) [reply client/cody-jetbrains/src/main/java/com/sourcegraph/cody/config/NotificationActivity.java:18]: Whoa, nice catch, an absolute copy&paste mistake.

File: client/cody-jetbrains/src/main/resources/META-INF/plugin.xml [hunk 3]
```diff
@@ -6,16 +6,35 @@
...
 
     <extensions defaultExtensionNs="com.intellij">
+        <projectService serviceImplementation="com.sourcegraph.cody.config.CodyProjectService"/>
+        <projectService serviceImplementation="com.sourcegraph.cody.config.SettingsChangeListener"/>
+        <applicationService serviceImplementation="com.sourcegraph.cody.config.CodyApplicationService"/>
+        <projectConfigurable
+            parentId="tools"
+            instance="com.sourcegraph.cody.config.ProjectSettingsConfigurable"
+            id="com.sourcegraph.cody.config.ProjectSettingsConfigurable"
+            displayName="Cody by Sourcegraph (Project Settings)"
```
[the remark this answers is not in the document]
Comment (vdavid) [reply client/cody-jetbrains/src/main/resources/META-INF/plugin.xml:16]: I don't think so. JetBrains usually uses a more complex solution that involves a dropdown at each settings page.
I've seen [this](https://plugins.jetbrains.com/plugin/274-jformdesigner) popular plugin use "JFormDesigner" and  "JFormDesigner (Project)" and copied their pattern. But "Cody" is much shorter than "JFormDesigner" so I actually added the longer but more precise "(Project Settings)" suffix.
One is not much better than the other, I guess, but we already had this two-leveled settings in the Sourcegraph plugin, and I've copied it to retain this functionality at a low cost. We can move things around later if our product designers have a better idea.

File: client/cody-jetbrains/src/main/java/com/sourcegraph/cody/telemetry/GraphQlLogger.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+import java.io.IOException;
+import java.util.function.Consumer;
+
+public class GraphQlLogger {
+    private static final Logger logger = Logger.getInstance(GraphQlLogger.class);
+
+    public static void logInstallEvent(Project project, Consumer<Boolean> callback) {
+        String anonymousUserId = ConfigUtil.getAnonymousUserId();
+        if (anonymousUserId != null) {
+            Event event = new Event("IDEInstalled", anonymousUserId, ConfigUtil.getSourcegraphUrl(project), null, null);
```
[the remark this answers is not in the document]
Comment (vdavid) [reply client/cody-jetbrains/src/main/java/com/sourcegraph/cody/telemetry/GraphQlLogger.java:22]: Well, tough call: we use `CodyInstalled` [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/cody/src/event-logger.ts?L48) with no referrer for the Cody VS Code extension, and we log no uninstall event there.
I think "IDEInstalled" with a different referrer is similarly good for tracking, but I've switched to "CodyInstalled" and "CodyUninstalled" now to make it extra explicit. Thanks for pointing it out!

File: client/cody-jetbrains/src/main/java/com/sourcegraph/cody/config/SettingsComponent.java [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+    @NotNull
+    private JPanel createOtherSettingsPanel() {
+        JBLabel codebaseLabel = new JBLabel("Codebase:");
+        codebaseTextField = new JBTextField();
+        //noinspection DialogTitleCapitalization
+        codebaseTextField.getEmptyText().setText("github.com/sourcegraph/sourcegraph");
+
+        // Always disabled for now
+        areChatPredictionsEnabledCheckBox = new JBCheckBox("Experimental: Chat predictions");
+        areChatPredictionsEnabledCheckBox.setEnabled(false);
```
[the remark this answers is not in the document]
Comment (vdavid) [reply client/cody-jetbrains/src/main/java/com/sourcegraph/cody/config/SettingsComponent.java:333]: Okay, commented it out. I have mixed feelings about it, too, but I wanted to leave in as a pointer for extensibility.

File: client/cody-jetbrains/src/main/java/com/sourcegraph/cody/config/SettingsConfigurableHelper.java (hunk 2 above)
Comment (vdavid) [reply client/cody-jetbrains/src/main/java/com/sourcegraph/cody/config/SettingsConfigurableHelper.java:21]: Yeah, I know. Considered breaking it up but it wouldn't be much nicer. Probably extracting this to its own class would be the Java way, but that's also 😭

**Pull request merged**